daemon/k8s_watcher.go

```diff
@@ -1449,6 +1449,32 @@ func getUpdatedCNPFromStore(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNetw
 	return serverRule, nil
 }
 
+func (d *Daemon) recalculatePolicyFromCNPUpdate(ciliumV2Store cache.Store, cnp *cilium_v2.CiliumNetworkPolicy) {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.CiliumNetworkPolicyName: cnp.ObjectMeta.Name,
+		logfields.K8sAPIVersion:           cnp.TypeMeta.APIVersion,
+		logfields.K8sNamespace:            cnp.ObjectMeta.Namespace,
+	})
+
+	scopedLog.Infof("updating node status due to annotations-only change to CiliumNetworkPolicy")
+
+	ctrlName := cnp.GetControllerName()
+
+	// Get the status for this node. If the status for this node hasn't been
+	// pushed to Kubernetes yet, the update will fail due to the resource being
+	// out of date; the controller will fail, and retry again later.
+	nodeName := node.GetName()
+	cnpns := cnp.GetPolicyStatus(nodeName)
+
+	k8sCM.UpdateController(ctrlName,
+		controller.ControllerParams{
+			DoFunc: func() error {
+				return cnpNodeStatusController(ciliumV2Store, cnp, cnpns.Revision, scopedLog, nil)
```

Comment:
I'm not sure I understand yet how this can work.

1. When the policy is initial added, we get a CNP without any Status fields from k8s. It will be stored in the store, we get a pointer to it.
2. We add it to the repository and get the revision for the CNP
3. We invoke the controller and run `cnpNodeStatusController()`. This is done async.
4. `cnpNodeStatusController()` will block until all endpoints have been recompiled.
5. In the meantime, an annotation update comes in, no spec change.
6. We fetch the CNP from the store, it does not yet have the Revision because the initial controller is still blocked and waiting for the endpoints to be regenerated.
7. Hence, the Revision field in the CNP will be zero.
8. We cancel the initial controller and restart it with `rev=0`. It will immediately set the Enforcing flag because revision 0 is clearly already implemented even though endpoints might still not be ready.

If you want this to work then you have to modify the Revison field of the CNP in the store before starting the initial controller to make sure it is updated and available when an update for that same CNP is processed.

Reply:
@tgraf 

Can't we modify the `UpdateController` to only update a controller once the previous controller is completed? This could be done by keeping a queue inside the controller. This would fix the annotation's bug this PR is fixing.

If the policy spec is actually modified, we would had a "force" field that would force the `UpdateController` to update the controller disregarding if the previous controller is still running or not (basically the current `UpdateController` behavior) and clean up the queue of controllers.

Reply:
I personally really don't want to modify the controller framework right now, especially since I view this as a release blocker and should be backported to v1.2 immediately before its release. There is a lot of critical code that depends on the current functionality, and I am afraid that even if we add testing for it, that we may reveal some odd behavior by changing the framework right now.

This is why I wanted to `TriggerPolicyUpdates` upon annotation update.

Reply:
To iterate upon what @tgraf proposed earlier - what if, instead of caching `CiliumNetworkPolicyNodeStatus` locally (which can be updated), a cache of just <namespace/name> --> revision at import time (which will always be the same per the definition of `Revision` in `CiliumNetworkPolicyNodeStatus` is used? 

This will be useful, because the calls to `addCiliumNetworkPolicyV2` and `updateCiliumNetworkPolicyV2` use the same `functionQueue` for enqueuing events. In a `functionQueue`, the functions in the queue are ran synchronously. This *ensures* no race-y behavior, because if a call to `updateCiliumnetworkPolicyV2` is added to the `functionQeue` for `CiliumNetworkPolicy`, then it is guaranteed that the function added before (`addCiliumNetworkPolicyV2 `) will come to completion before `updateCiliumnetworkPolicyV2` is called. This means that when `updateCiliumnetworkPolicyV2` is called, the local cache referenced above, will always have the revision for the `CiliumNetworkPolicyNodeStatus`.

Reply:
The proposed behavior above will indeed work, because the cache is updated synchronously in `addCiliumNetworkPolicyV2` before `updateCiliumNetworkPolicyV2` is called. I believe that this is the lowest-risk way of fixing this in the scope that is available to us.

Reply:
I added the update to this cnp --> revision cache in `addCiliumNetworkPolicyV2`, and then a sleep for ten seconds to `addCiliumNetworkPolicyV2` to confirm that this raciness will not occur - this allowed for me to add a policy, then update it shortly after and see if the update occurred before the add. Since the cache is not updated in the controller, the cache was correctly populated for `updateCiliumNetworkPolicyV2` to retrieve it:

For debugging purposes, I added a UUID to each function queue, then logged the UUID of the function queue for CNPs only to reduce the amount of logs to trawl through (this is where `e3ab9a5a-a006-11e8-ad63-0800278b583` comes from). I also logged a UUID for each sleep log from `addCiliumNetworkPolicy`, and a log for each operation on the new cache. The ordering shows that the behavior works as expected:

```
$ grep -r "sleeping for 10 seconds to create racey behavior\|done sleeping for 10 seconds to create racey behavior\|ruleRevisionCache\|functionQueue e3ab9a5a-a006-11e8-ad63-0800278b583\|serCNPs" /var/log/syslog
Aug 14 21:13:28 vagrant cilium-agent[15572]: level=info msg="serCNPs UUID: e3ab9a5a-a006-11e8-ad63-0800278b5835" subsys=daemon
Aug 14 21:13:28 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func e3b1d679-a006-11e8-ad63-0800278b5835"
Aug 14 21:13:28 vagrant cilium-agent[15572]: level=warning msg="sleeping for 10 seconds to create racey behavior: e3b1ea92-a006-11e8-ad63-0800278b5835" subsys=daemon
Aug 14 21:13:38 vagrant cilium-agent[15572]: level=warning msg="done sleeping for 10 seconds to create racey behavior: e3b1ea92-a006-11e8-ad63-0800278b5835" subsys=daemon
Aug 14 21:13:38 vagrant cilium-agent[15572]: level=info msg="ruleRevisionCache.upsert: default/cnp-default-deny-egress --> 2" subsys=daemon
Aug 14 21:13:38 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func e3b1d679-a006-11e8-ad63-0800278b5835"
Aug 14 21:13:38 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func e9aab862-a006-11e8-ad63-0800278b5835"
Aug 14 21:13:38 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func e9aab862-a006-11e8-ad63-0800278b5835"
Aug 14 21:14:09 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func fbddf138-a006-11e8-ad63-0800278b5835"
Aug 14 21:14:09 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func fbddf138-a006-11e8-ad63-0800278b5835"
Aug 14 21:14:09 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func fbf18e58-a006-11e8-ad63-0800278b5835"
Aug 14 21:14:09 vagrant cilium-agent[15572]: level=info msg="ruleRevisionCache.delete: default/cnp-default-deny-egress" subsys=daemon
Aug 14 21:14:09 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func fbf18e58-a006-11e8-ad63-0800278b5835"
Aug 14 21:14:23 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func 045c9df5-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:23 vagrant cilium-agent[15572]: level=warning msg="sleeping for 10 seconds to create racey behavior: 045ca3ce-a007-11e8-ad63-0800278b5835" subsys=daemon
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=warning msg="done sleeping for 10 seconds to create racey behavior: 045ca3ce-a007-11e8-ad63-0800278b5835" subsys=daemon
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="ruleRevisionCache.upsert: default/cnp-default-deny-egress --> 4" subsys=daemon
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func 045c9df5-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func 0a529520-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="ruleRevisionCache.get: default/cnp-default-deny-egress --> 4, true" subsys=daemon
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func 0a529520-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func 0a53bd55-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func 0a53bd55-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: running func 0a552d62-a007-11e8-ad63-0800278b5835"
Aug 14 21:14:33 vagrant cilium-agent[15572]: level=info msg="functionQueue e3ab9a5a-a006-11e8-ad63-0800278b5835: done running func 0a552d62-a007-11e8-ad63-0800278b5835"
```

Reply:
Sounds good